src/CommonDBTM.php

```diff
@@ -667,32 +667,31 @@ public function updateInDB($updates, $oldvalues = [])
     {
         global $DB;
 
+        $tobeupdated = [];
         foreach ($updates as $field) {
             if (isset($this->fields[$field])) {
-                $DB->update(
-                    $this->getTable(),
-                    [$field => $this->fields[$field]],
-                    ['id' => $this->fields['id']]
-                );
-                if ($DB->affectedRows() == 0) {
-                    if (isset($oldvalues[$field])) {
-                        unset($oldvalues[$field]);
-                    }
-                }
-            } else {
-               // Clean oldvalues
-                if (isset($oldvalues[$field])) {
+                if (isset($oldvalues[$field]) && $this->fields[$field] == $oldvalues[$field]) {
                     unset($oldvalues[$field]);
                 }
+                $tobeupdated[$field] = $this->fields[$field];
+            } else {
+                // Clean oldvalues
+                unset($oldvalues[$field]);
             }
         }
+        $DB->update(
+            $this->getTable(),
+            $tobeupdated,
+            ['id' => $this->fields['id']]
+        );
+        $affected_rows = $DB->affectedRows();
 
-        if (count($oldvalues)) {
+        if (count($oldvalues) && $affected_rows > 0) {
             Log::constructHistory($this, $oldvalues, $this->fields);
             $this->getFromDB($this->fields[$this->getIndexName()]);
         }
 
-        return true;
+        return ($affected_rows >= 0);
```

Comment:
Having no affected row does not mean there is a failure. It just mean that the piece of code that call `updateInDB` is not optimized and did not checked that the new value is different than the existing value.

So, for me, we should consider that the operation was a success (e.g. property in DB has now the same value as the one passed in `$this->fields`), even if operation was useless.

```suggestion
        return true;
```

Reply:
Having affectrows < 0 is an error as far as I've seen

Reply:
Sorry, I misread the condition. As it will return true if `$affected_rows == 0`, it is ok for me.

Note that we will probably never have a `-1` value here as it would ptobably trigger the `$result === false` case.